renderdoc/driver/gl/gl_pixelhistory.cpp

```diff
@@ -276,13 +383,226 @@ void QueryFailedTests(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
     history[i].backfaceCulled = failedTest == OpenGLTest::FaceCulling;
   }
 }
-};
+
+void QueryPerFragmentValues(WrappedOpenGL *driver, GLReplay *replay,
+                            GLPixelHistoryResources &resources, const EventUsage &modEvent, int x,
+                            int y, rdcarray<PixelModification> &history,
+                            const std::map<uint32_t, uint32_t> &eventFragments,
+                            PerFragmentQueryType queryType)
+{
+  driver->ReplayLog(0, modEvent.eventId - 1, eReplay_Full);
+  driver->glBindFramebuffer(eGL_FRAMEBUFFER, resources.frameBuffer);
+  driver->glReadBuffer(eGL_COLOR_ATTACHMENT0);
+
+  auto it = eventFragments.find(modEvent.eventId);
+  uint32_t numFragments = (it != eventFragments.end()) ? it->second : 0;
+
+  driver->glEnable(eGL_SCISSOR_TEST);
+  driver->glScissor(x, y, 1, 1);
+
+  driver->glClearStencil(0);
+
+  if(queryType == PerFragmentQueryType::ShaderOut || queryType == PerFragmentQueryType::PrimitiveId)
+  {
+    driver->glEnable(eGL_DEPTH_TEST);
+    driver->glDepthFunc(eGL_ALWAYS);
+    driver->glDepthMask(GL_TRUE);
+    driver->glDisable(eGL_BLEND);
+  }
+
+  driver->glStencilMask(0xff);
+  driver->glStencilOp(eGL_INCR, eGL_INCR, eGL_INCR);
+  driver->glEnable(eGL_STENCIL_TEST);
+
+  PixelModification referenceHistory;
+  referenceHistory.eventId = modEvent.eventId;
+
+  auto historyIndex =
+      std::lower_bound(history.begin(), history.end(), referenceHistory,
+                       [](const PixelModification &h1, const PixelModification &h2) -> bool {
+                         return h1.eventId < h2.eventId;
+                       });
+  RDCASSERT(historyIndex != history.end());
+
+  if(queryType == PerFragmentQueryType::PostMod)
+  {
+    if(historyIndex != history.begin())
+    {
+      --historyIndex;
+
+      // Because we are replaying this draw into our own framebuffer
+      // We need to set the values to the ones that they would be
+      // before this draw call.
+      driver->glClearColor(
+          historyIndex->postMod.col.floatValue[0], historyIndex->postMod.col.floatValue[1],
+          historyIndex->postMod.col.floatValue[2], historyIndex->postMod.col.floatValue[3]);
+
+      GLboolean depthWriteMask = false;
+      driver->glGetBooleanv(eGL_DEPTH_WRITEMASK, &depthWriteMask);
+      driver->glDepthMask(true);
+      driver->glClearDepth(historyIndex->postMod.depth);
+
+      driver->glClear(eGL_COLOR_BUFFER_BIT | eGL_DEPTH_BUFFER_BIT);
+
+      driver->glDepthMask(depthWriteMask);
+      ++historyIndex;
+    }
+  }
+  else if(queryType == PerFragmentQueryType::PrimitiveId)
+  {
+    // we expect this value to be overwritten by the primitive id shader. if not
+    // it will cause an assertion that all color values are the same to fail
+    driver->glClearColor(0.84f, 0.17f, 0.2f, 0.49f);
+    driver->glClear(eGL_COLOR_BUFFER_BIT);
+
+    GLint currentProgram;
+
+    driver->glGetIntegerv(eGL_CURRENT_PROGRAM, &currentProgram);
```

Comment:
This program isn't restored afterwards. Currently it doesn't matter that the state is polluted but this can easily cause problems in future if the state should remain undisturbed after the work.